### ISSUE TITLE
ITI-68: incompatibility with the FHIR Binary read operation

### ISSUE THREAD

**Issue by qligier** [IHE/ITI.MHD#233]
**Section Number** [2:3.68.4.1.2 Message Semantics](https://profiles.ihe.net/ITI/MHD/ITI-68.html#2368412-message-semantics)

**Issue**
At the European Connectathon, we have seen multiple vendors using a regular FHIR Binary read URL as ITI-68 endpoint (e.g. `DocumentReference.content.attachment.url = http://example.com/fhir/Binary/a54ef236-7e13-48be-951c-1ab7a6a2611f`).

This seems to be non-compliant with the ITI-68 specifications in some edge cases:

- If the stored document is anything but a FHIR resource, and the ITI-68 request contains `Accept: application/fhir+xml` (or JSON)
MHD should fail with a 406 Not Acceptable, or convert the document into a FHIR resource (document Bundle)
FHIR should return the FHIR Binary.
- If the stored document is a FHIR resource, and the ITI-68 request contains `Accept: application/fhir+xml` (or JSON)
MHD should return the document content (a FHIR resource).
FHIR should return the FHIR Binary.

One might argue that transforming the stored document into a FHIR Binary is allowed by [2:3.68.4.1.2](https://profiles.ihe.net/ITI/MHD/ITI-68.html#2368412-message-semantics), so the first example is compliant. But the second one definitively isn't.

Is that incompatibility between both specifications already known?

**Proposed Change**
If it is a deliberate choice, could an explicit warning be added to the ITI-68 specification to inform vendors the FHIR Binary read is not suitable as ITI-68 endpoint?
If it is not a deliberate choice, should we discuss whether ITI-68 specifications should be updated to match the FHIR Binary read operation? The main argument in favour that I can think of is the simplicity of implementation (and unification of specifications). 

**Priority**:
Medium

@oliveregger FYI

**Comment by JohnMoehrke** [comment]
The Document Responder is completely empowered to use whatever URL format they want. The Document Consumer must not expect any specific URL format, they must just do a GET. The Document Responder can use a URL format that appears to look like a Binary endpoint, it is their choice.  The Document Responder can invent their own URL encoding, including using url parameters.

Note in MHDS, where the Document Registry is persisting the documents, these URLs will be Binary endpoints.

see https://profiles.ihe.net/ITI/MHD/ITI-67.html#236742211-document-location

This does setup the situation you describe, and that situation is fully expected. If the Document Consumer really does put FHIR in the accept header higher up than the mime-type of the document, then they ARE ASKING for a FHIR encoding as higher priority, and thus getting a FHIR Binary should be expected.

This is explained - https://profiles.ihe.net/ITI/MHD/ITI-68.html#2368412-message-semantics

Meaning that the Document Consumer must understand what they have placed into the http accept header, as it could change the mime-type of the content returned.

I speak to the benefits of using http negotiate on the ITI-68 in my tutorials.

**Comment by JohnMoehrke** [comment]
committee review:
It is unclear what problem you are trying to describe. As I have commented above, ITI-68 has few requirements. ITI-68 is compatible with a FHIR Binary endpoint, but  ITI-68 does not require a FHIR Binary endpoint. The folder `/Binary/` is not exclusive to FHIR Binary definition. The http Accept header allows the server to return the content they have.

Please explain the exact problem. Specifically what has happened that has caused what bad behavior. What text in the MHD specification is a problem, and how would you recommend it be changed.

**Comment by qligier** [comment]
We have seen different implementations of Document Responders and there was disagreement what the expected response should be for ITI-68 when using an Accept Header with application/fhir+json / application/fhir+xml.

Let’s look at examples to illustrate the issue. Let’s take two different implementations, one based on a FHIR server with Binary endpoint for the files and one custom implementation just using a storage.

**1\. Example;  If the stored document is a FHIR resource (e.g. Bundle in this example) in XML representation, we get the following result**

| MIME-type of the stored document: application/fhir+xml | FHIR implementation<br>`GET /Binary/{id}` | Custom implementation<br>`GET /custom/path` |
| :---- | :---- | :---- |
| No Accept header | The XML Bundle is returned | The XML Bundle is returned |
| Accept \= application/fhir+xml | The XML **Binary** resource containing the Bundle is returned [link](https://hl7.org/fhir/R4/binary.html\#:\~:text=the%20RESTful%20API%3A-,When%20a%20read%20request%20is%20made%20with%20a%20FHIR%20type%20in%20the%20Accept%20header%20(e.g.%20%22application/fhir%2Bxml%22%20or%20%22application/fhir%2Bjson%22)%20the%20Binary%20resource%20is%20returned%20in%20the%20requested%20FHIR%20format.%20This%20applies%20even%20when%20the%20binary%20data%20itself%20has%20a%20FHIR%20mime%20type,-When%20a%20read) | The XML **Bundle** is returned [link](https://profiles.ihe.net/ITI/MHD/ITI-68.html\#2368412-message-semantics) |
| Accept \= application/fhir+json | The JSON **Binary** resource containing the Bundle is returned [link](https://hl7.org/fhir/R4/binary.html\#:\~:text=the%20RESTful%20API%3A-,When%20a%20read%20request%20is%20made%20with%20a%20FHIR%20type%20in%20the%20Accept%20header%20(e.g.%20%22application/fhir%2Bxml%22%20or%20%22application/fhir%2Bjson%22)%20the%20Binary%20resource%20is%20returned%20in%20the%20requested%20FHIR%20format.%20This%20applies%20even%20when%20the%20binary%20data%20itself%20has%20a%20FHIR%20mime%20type,-When%20a%20read) | Either 406 not Acceptable, or the JSON **Bundle** is returned [link](https://profiles.ihe.net/ITI/MHD/ITI-68.html\#2368412-message-semantics) |

The issue lies in the difference of responses when using the *Accept:* *application/fhir+xml* or *Accept:* *application/fhir+json* header for a stored document with a content-type of *application/fhir+xml*  or *application/fhir+json* : the returned content is dependent on the Document Responder implementation, if it uses the FHIR Binary you get a FHIR Binary Resource back, otherwise you get a FHIR Bundle resource back. This is a problem for the Document Consumer as we have seen it at the connectathon. The Document Consumer has no way of knowing the behaviour and the Document Responder cannot declare which variation it supports, and we have seen both implementations in Document Responders and the consumer struggling what to expect. We also saw a Document Consumer who always used the Accept header with *application/fhir+json* to specifically request the Binary resource for all documents.

**2\. Example: If the stored document is a PDF, we get the following result:**

| MIME-type of the stored document: application/pdf | FHIR implementation<br>`GET /Binary/{id}` | Custom implementation<br>`GET /custom/path` |
| :---- | :---- | :---- |
| No Accept header | The PDF is returned | The PDF is returned |
| Accept \= application/pdf | The PDF is returned | The PDF is returned |
| Accept \= application/fhir+json | The JSON **Binary** resource containing the PDF is returned [link](https://hl7.org/fhir/R4/binary.html\#:\~:text=the%20RESTful%20API%3A-,When%20a%20read%20request%20is%20made%20with%20a%20FHIR%20type%20in%20the%20Accept%20header%20(e.g.%20%22application/fhir%2Bxml%22%20or%20%22application/fhir%2Bjson%22)%20the%20Binary%20resource%20is%20returned%20in%20the%20requested%20FHIR%20format.%20This%20applies%20even%20when%20the%20binary%20data%20itself%20has%20a%20FHIR%20mime%20type,-When%20a%20read) | Either 406 Not Acceptable, or a JSON resource is returned [Link](https://profiles.ihe.net/ITI/MHD/ITI-68.html\#2368412-message-semantics) |

This is also a similar issue,  as most of the Document Responders which do not provide the Document via a FHIR API return a 406, which is correct, but the Document Consumer does not know which behaviour it has to expect.

From the discussion FHIR Binary endpoint is allowed in ITI-68, we suggest adding a sentence such as:

> **Note:** If the document is encoded as a FHIR resource (`application/fhir+xml` or `application/fhir+json`), and the Document Consumer includes in the request an Accept header with the original MIME type, then the Document Responder may return a FHIR Binary resource containing the original document content.

In addition, the sentence:

> _For example, indicating application/fhir+json could result in the response from the Document Responder being a json FHIR Bundle of type document with all the content encoded as FHIR resources._

will never be realisable if the Document Responder is using a Binary endpoint, we suggest removing this sentence, or changing it for another example.

**Comment by JohnMoehrke** [comment]
From the description of the problem, I don't see how your recommended additional sentence helps.

The problem statement indicates that there are clients that are not following the http specification. If they were to simply follow that specification they would understand that 406 is a legitimate error code for when the server can't satisfy the accept header limits.   

ITI-68 is purely http. It is not constrained to FHIR Binary.  The references section should be changed to state this, and not mention FHIR. I am not sure what is confusing people into thinking that ITI-68 is describing FHIR Binary, it is not. It is simply describing http GET.

the FHIR Attachment.url is also just a url, it is not constrained to FHIR Binary.

Emphasizing these points seem more useful points?

**Comment by JohnMoehrke** [comment]
Note that if the client indicates in accept that they understand FHIR, then they are indicating that they understand FHIR. Thus they would be able to distinguish these various responses by looking at the response mime-type to understand it as a FHIR response vs something else.  Emphasis that this is simply http

**Comment by oliveregger** [comment]
The problem statement indicates that we have different behaviour in ITI-65 response if the stored document has a mime-type from FHIR depending on the implementation of the Document Responderwhere a) it is a plain url or b) is also to be allowed to point to a FHIR Binary endpoint. This has nothing to do with the client / 406 or correct not following the http specification.  Why should we not put on a note that this situation can occur when we see the different behaviours in implementations? 
@JohnMoehrke do you disagree that this different behaviour exists as it is pointed out?

>Note: If the document is encoded as a FHIR resource (application/fhir+xml or application/fhir+json), and the Document Consumer includes in the request an Accept header with the original MIME type, then the Document Responder may return a FHIR Binary resource containing the original document content.

**Comment by oliveregger** [comment]
the results are not consistent with http get. I put out two exactly http get calls with an accept header (which is allowed by http get), but depending on the document responder implementation I get a structural complete different results, this is not interoperability ...

we dont need to  emphasize that ITI-68 Binary, but we should be able to detect it as a consumer if it is used in iti-68. Why is the capabilitystatement of the MHD Document Responder indicating [Binary READ](https://profiles.ihe.net/ITI/MHD/CapabilityStatement-IHE.MHD.DocumentResponder.html)? would that be a way to point out the differentiation?

**Comment by JohnMoehrke** [comment]
what are the deviations from http get that you have seen? Those are what we should focus on.

The capabilityStatement likely should be corrected. That one is not wrong for a system that did choose to implement ITI-68 with Binary endpoint. But ITI-68 does not require Binary endpoint.

**Comment by qligier** [comment]
The issue is that the ITI-68 specification does not detail enough what will be returned by the Document Responder to the Document Consumer when using an Accept header. Anything could be returned Even when the consumer asks for a document of the same MIME type as the original content, the responder could return anything, as long as the MIME type is correct. That is valid according to the HTTP specifications, but not for an IHE transaction because it is simply not interoperable.
If the consumer has to ask the responder for implementation details just to be able to retrieve a document, that is not interoperability, that's a proprietary API.

**Comment by oliveregger** [comment]
> if the client indicates in he accept headers that they understand the mime-type indicated in attachment.contentType, then I would expect that content to be returned.
> 
> What is not clear?

That we have undetermined behaviour in the response as we indicate in the example 1. HTTP negotiation is completely the same for the client but the response is different.

> I am not against being more specific in ITI-68; but I will not accept that ITI-68 is a FHIR Binary endpoint. That is too limiting to both the server and client. http negotiation is sufficient.

How are you meaning this? Neither @qligier nor myself are advocating that it SHALL be FHIR Binary endpoint. But is your sentence meant that it SHALLT NOT be a FHIR Binary endpoint?

**Comment by JohnMoehrke** [comment]
please put forward a proposed change.

**Comment by JohnMoehrke** [comment]
This was discussed at IHE ITI face-to-face. There might be interest in making a clarification. The committee recommends that a PR be submitted that indicates how the specification should change to add the clarity.

**Comment by oliveregger** [comment]
see also discussion in zulip https://chat.fhir.org/#narrow/channel/179223-ihe/topic/iti-68

**Comment by qligier** [comment]
Here is a proposed solution, as it was discussed on the Zulip thread:

The goal is to clarify the expected Document Responder behaviour regarding the response content, depending on the Accept header value or absence.

- When the Document Consumer does not set the Accept header, the Document Responder SHALL return the original content (the binary data as it was published).
- When the Document Consumer does set the Accept header, the Document Responder MAY return the original content or MAY return a content transformed from the original content (e.g. wrapping it in a FHIR Binary resource), or MAY return an error (406 Not Acceptable). The specification offers no guarantee to the Consumer, who has to get knowledge from the Responder (e.g. are you a FHIR compliant server, which transformations do you support, etc.).

The default behaviour for the Document Consumer would then be not to set the Accept header, except if they have some knowledge of the Document Responder implementation.

A note should be added, explaining that due to the predominance of FHIR servers used as Document Responder, Document Consumers should be aware of the special handling of the MIME types "application/fhir+xml" and "application/fhir+json".

This solution is coherent with the intent of [RFC 7231](https://datatracker.ietf.org/doc/html/rfc7231#section-5.3.2) and the [FHIR specification for Binary read](https://www.hl7.org/fhir/R4/binary.html#rest):

> When a read request is made with a FHIR type in the Accept header (e.g. "application/fhir+xml" or "application/fhir+json") the Binary resource is returned in the requested FHIR format. This applies even when the binary data itself has a FHIR mime type.

> due to the way the web infrastructure works, it is not possible to make blanket rules about the relationship between the "Accept" field in the HTTP request, and the return type, which is why there is no hard rule about this. However, the intent is that unless specifically requested, the FHIR XML/JSON representation is not returned.

There is currently a [small issue in HAPI](https://github.com/hapifhir/hapi-fhir/issues/6630) we saw when testing this behaviour, it returns the contained resource instead of the Binary if `Binary.contentType` matches the Accept header value when reading. Otherwise, it conforms to the FHIR spec.

**Comment by JohnMoehrke** [comment]
Please propose a CP with these recommendations.

**Comment by heathfrankel** [comment]
Just correcting my comment above. The FHIR standard says

> When the read request has some other type in the Accept header, then the content should be returned with the content type stated in the resource in the Content-Type header. E.g. if the content type in the resource is "application/pdf", then the content should be returned as a PDF directly.

> the intent is that unless specifically requested, the FHIR XML/JSON representation is not returned

Therefore, only when a FHIR mime type is provided in the Accept header the Binary resource is returned, otherwise the content is returned with the content type specified in the Binary.contentType, including when the Accept header is not provided as you suggested above.

In this case, where no Accept header is provided in the request, and the DocumentReference references a Binary endpoint, it would be valid to return the document Bundle. 
Although the CapabilityStatement I reference above indicates the Binary resource is supported, it is only indicating the resource endpoint is supported, not that the response content shall be a Binary resource. In fact when there is no Accept header, it is behaving as expected and returning the document Bundle as the content of the Binary resource, which is the desirable outcome given we don't want to have to extract the content from a Binary resource as a secondary step.

In summary, I still don't think this issue is an issue, and no change is required except perhaps to indicate that an Accept header with a FHIR mime type should not be provided in the ITI-68 request otherwise they will get an unexpected response of a Binary resource rather than its content as intended.

**Comment by qligier** [comment]
Hi @heathfrankel!

> your proposed change above actually contradicts the FHIR specification, which as you quoted states

It doesn't violate the FHIR specs. You forgot the first part of that sentence, that says:

> When a read request is made with a FHIR type in the Accept header (e.g. "application/fhir+xml" or "application/fhir+json") the Binary resource is returned [...]

I.e. it applies if and only if an _Accept_ header is specified. My proposal says:

> When the Document Consumer does set the Accept header, the Document Responder MAY return the original content or MAY return a content transformed from the original content [...]

I don't see why it would violate the FHIR specification. Whether you consider the Binary resource as the original content (as you seem to do) or as a transformed content (as I do), it would be conforming.

> I would also argue that by not providing an Accept header, you are requesting the resource at that location in its default representation, i.e. a Binary resource in the server's default FHIR format.

I see your point, but that's debatable. Is the default representation the Binary resource, or the content in `Binary.data`? Coming from the XDS world, I would argue for the latter; but if we settled for the former, it would simplify that transaction: the client would do a FHIR Binary read, and the default representation would be the Binary resource.
But we seem to agree on the fact that without an Accept header, the original content should be returned.

> It is only when an alternate Accept value that matches the Binary.contentType where it returns the original binary content

That's specifically what's NOT happening in the FHIR specification for the special case where the Accept header is a FHIR media type, and is equal to `Binary.contentType`: in that case, the Binary resource is served instead of the original content (that may not be binary).

> In other words, HAPI is actually behaving as expected

It's not behaving according to the FHIR specification. I'm not sure if it's a simple bug, or if that behaviour was desired for a compatibility or convenience reason. The issue has all details showing it's not conforming: https://github.com/hapifhir/hapi-fhir/issues/6630

> Also note that the [IHE DocumentResponder CapabilityStatement](https://profiles.ihe.net/ITI/MHD/CapabilityStatement-IHE.MHD.DocumentResponder.html#resourcesCap1) indicates that it supports the Binary endpoint for this ITI-68 transaction.

John said ITI-68 is not a Binary read but a standard HTTP GET, so the CapabilityStatement is irrelevant here. (To be clear, constraining ITI-68 to a Binary read would be a good solution to that issue I think).

> Therefore, I think if I understand your original documented issue above

I think you missed the issue by focusing on FHIR. The issue I see is that a Document Consumer doing an ITI-68 with an Accept header has no way of knowing if they are being served the original content or a transformed one. That makes interoperability impossible, because you need to know details about the Document Responder implementation to know what you got in response.

Simple example: you make an ITI-68 request with `Accept: application/fhir+json`. You get a Binary resource in response. What do you do with it?

- From what you said, you would extract the `Binary.data`, and read it as the original content. That would happen if the Document Responder is a FHIR server.
- Now, what if the Document Responder is a HAPI server? The Binary resource you got is not the wrapping Binary resource, but the one that was published (i.e. as if you published that Binary resource with an ITI-41).
- If the Document Responder is not a FHIR server, you are probably also being served the original content, and you shouldn't extract the `Binary.data`

So, as a Document Consumer, you can't process that response without knowing details about the Document Responder implementation, what is impossible programmatically in MHD, and so prevents interoperability.

I hope it's clearer with this explanation.

**Comment by heathfrankel** [comment]
@qligier I think we crossed over messages. See above, I corrected myself on what should happen when no Accept header is provided.

> constraining ITI-68 to a Binary read would be a good solution to that issue I think

Because of the general nature of ITI-68 this is not possible from what I can tell. I think it can only be assumed if ITI-65 FHIR Document Publish Option is grouped with MHDS Document Registry (see https://profiles.ihe.net/ITI/MHD/ITI-65.html#23654132-grouping-with-mhds-implementation-guide-profile). I am still trying to get my head around all the possible groupings, options and transactions.

**Comment by qligier** [comment]
_Follow-up response to your second message; our messages crossed_

> Therefore, only when a FHIR mime type is provided in the Accept header the Binary resource is returned, otherwise the content is returned with the content type specified in the Binary.contentType

In the case where the Accept header is not set, yes. Otherwise, it's unspecified by the FHIR specification.

> In this case, where no Accept header is provided in the request, and the DocumentReference references a Binary endpoint, it would be valid to return the document Bundle.

Valid according to the FHIR specification? Yes. But again, it's not relevant because the ITI-68 doesn't require serving the original content when the Accept header is not specified. So it's a possibility, but a Document Responder MAY behave otherwise.

> Although the CapabilityStatement I reference above indicates the Binary resource is supported, it is only indicating the resource endpoint is supported, not that the response content shall be a Binary resource.

Yes, that's what the FHIR specification says. But again, ITI-68 is not a FHIR API and the CapabilityStatement is not relevant. A Document Responder doesn't need to be a FHIR server.

> In fact when there is no Accept header, it is behaving as expected and returning the document Bundle as the content of the Binary resource, which is the desirable outcome given we don't want to have to extract the content from a Binary resource as a secondary step.

It's not behaving as expected, because the ITI-68 specification doesn't set any expectation. It just says the Document Consumer may or may not set the Accept header, and the Document Responder may or may not modify the content.

> In summary, I still don't think this issue is an issue

With the details provided in my two responses, I hope to provide enough information to deny that claim.

> except perhaps to indicate that an Accept header with a FHIR mime type should not be provided in the ITI-68 request

Interesting possibility, I hadn't thought of that. It might be a partial solution, but it doesn't solve the issue completely. The ITI-68 specification still has to explicitly set expectations about the Document Responder response when no Accept header is provided, and when the Accept header is provided and is equal to the original media type (as conveyed in `Attachment.contentType`).

**Comment by qligier** [comment]
> If I am on the right track now

Yes, I think you captured the issue. It's not necessarily linked to FHIR, because two non-FHIR implementations could be non-interoperable while being ITI-18-conformant. But due to the number of implementers using a FHIR server, the FHIR rules are creating the most recurrent occurrences of this bug.

> Something else that needs to be tidied up

Indeed, I missed that part. If ITI-68 says 'SHALL provide a CapabilityStatement', and the CapabilityStatement says you SHALL support the Binary read, then it says a Document Responder SHALL implement a Binary read endpoint (and MAY implement another endpoint, like one similar to ITI-43). That should definitively be fixed. @JohnMoehrke?

**Comment by heathfrankel** [comment]
> The ITI-68 specification still has to explicitly set expectations about the Document Responder response when no Accept header is provided, and when the Accept header is provided and is equal to the original media type (as conveyed in Attachment.contentType).

Yes, I agree this will help. Especially the paragraph that is suggesting that the FHIR mime type will return a FHIR document Bundle, which it won't when grouped with a MHDS Document Registry (i.e. implemented using a FHIR Server).

**Comment by heathfrankel** [comment]
The CapabilityStatement does NOT say 'you SHALL support the Binary read'. 
It only says 'Supports read'

None of the resources have an interaction conformance requirement (yet).

**Comment by qligier** [comment]
It says 'Supports read' on the Binary resource, in a CapabilityStatement of type 'requirements'. I interpreted that as a 'SHALL implement', but I may be wrong here.

It seems like the expectations can be clarified, as per the [CapabilityStatement documentation](https://hl7.org/fhir/R4/capabilitystatement.html#requirements):

> CapabilityStatements of type "requirement" describe what capabilities are potentially relevant; additional documentation or extensions (see [capabilitystatement-expectation](https://hl7.org/fhir/R4/extension-capabilitystatement-expectation.html)) within the CapabilityStatement are expected to make more explicit statements of degree of expectation associated with each capability.

**Comment by heathfrankel** [comment]
See https://hl7.org.au/fhir/core/1.0.0/CapabilityStatement-au-core-responder.html#Patient1-11 for an example when explicit conformance requirements have been stated against an interaction, or resource itself.  
These are added using capabilitystatement-expectation extensions.

**Comment by JohnMoehrke** [comment]
I am having a hard time following the discussion. I think however we are all trying to get 

1. The best possible experience for a MHD Document Consumer - This actor should be able to get as close to exactly what it wants, and should not be confused by the result that it gets (I am not worried about a Document Consumer that says it handles FHIR, getting a Binary, as it clearly said it understands FHIR and thus it should clearly understand what a Binary or Bundle is).
2. The least customization for a MHDS Document Registry (or virtual repository as MHDS does not guarantee all documents are in the Registry.) Best case is that a non-customized FHIR server could be a MHDS Document Registry.
3. we need to support where MHD is the api to MHDS but also XDS, XCA. And on the ITI-65 side support for push. 
4. Deterministic and documented behaviors. This I think is the key problem that started this discussion as some servers were behaving differently and thus some clients were not expecting. The least we can do is document what will/must happen.

Thus, we are indeed balancing many factors. We can't fix it for one environment and break another. 

Let's fix whatever needs to be fixed. 

I was discussing this issue with Heath on zulip at the same time he was posting here.

A concept Heath and I talked about is that we might need to have special handling of MHD ITI-65 when it is backed by MHDS, such that if the content is a FHIR Document Bundle, that it is persisted in FHIR Document Bundle form, and not Binary. All other mime-types would be preserved as Binary.  Thus the Document Responder does not just return Binary but also Bundle (in special conditions); 

however we also discussed that the CS for Document Responder could be completely silent on Binary and Bundle; as it is really just responding to an http GET on a URL that it provided earlier in the DocumentReference.content.attachment.url. (MHDS Document Registry likely should say both, but not MHD Document Responder.

I do like the observation that FHIR Binary when the client does not provide any http negotiate would be returning the Binary.data unencoded; this seems key. And I think in the MHD->XDS; this would work as well. As http GET without negotiate does expect to get whatever format is default, which would be the mime-type in the DocumentReference

**Comment by qligier** [comment]
Something like this?

2. When the Document Consumer does set the Accept header:
   1.  If the Accept header value is one of the FHIR media types (`application/fhir+json`, `application/fhir+xml`), then the Document Responder SHALL return a Binary resource including the original content in `Binary.data`.
   2. Otherwise, if the Accept header value is equal to the one provided in the DocumentReference, the DocumentResponder SHALL return the original content (?).
   3. Otherwise, the Document Responder MAY transform the original content to fullfill the Document Consumer request, or MAY return a `406 Not Acceptable` error if it is unable to.

**Comment by slagesse-epic** [comment]
2.i disallows "dumb" content repositories that aren't FHIR servers from hosting the binaries, I think.
